### PR TITLE
Parse XCbuild stderr only if stdout is empty

### PR DESCRIPTION
**Pull request by elsh**
SwiftPM parses all stderr output from XCBuild that shouldn't lead to a build failure, but emits it as an error diagnostic, which does lead to it. Diagnostics are already available via the parseable JSON output on stdout, not textually via stderr.

### Motivation:
Resolves rdar://75751711 (SwiftPM is parsing _ALL_ stderr output from XCBuild into an error diagnostic that fails the build).

### Modifications:
This change parses XCBuild stderr and emits error only if stdout is empty. 

### Result:
If XCBuild stdout is not empty, stderr is now ignored, else it will throw a build error.

### CODE REVIEW

File: Sources/XCBuildSupport/XcodeBuildSystem.swift [hunk 1]
```diff
@@ -104,10 +104,13 @@ public final class XcodeBuildSystem: SPMBuildCore.BuildSystem {
         arguments += buildParameters.xcbuildFlags
 
         let delegate = createBuildDelegate()
+        
         let redirection: Process.OutputRedirection = .stream(stdout: delegate.parse(bytes:), stderr: { bytes in
-            self.diagnostics.emit(StringError(String(bytes: bytes, encoding: .utf8)!))
+            if let str = String(bytes: bytes, encoding: .utf8) {
```
Comment (abertelrud) [inline Sources/XCBuildSupport/XcodeBuildSystem.swift:109]: Also, is it the right thing to emit all of this output as warnings?  If I have an error that is emitted, would it show up as parseable diagnostics as well as on stderr, and thus show up as a warning as well as an error?

Comment (elsh) [reply Sources/XCBuildSupport/XcodeBuildSystem.swift:109]: That's what the report in the radar seems to suggest. @jakepetroules Could you confirm the above question?  Or would we need to filter and treat certain stderr as actual errors?

Comment (jakepetroules) [reply Sources/XCBuildSupport/XcodeBuildSystem.swift:109]: All XCBuild output is communicated as structured (JSON) data via stdout, if we emit anything via stderr it's unintentional and can be dropped.

**Comment by elsh** [comment]
@swift-ci smoke test

File: Sources/XCBuildSupport/XcodeBuildSystem.swift [hunk 2]
```diff
@@ -119,6 +116,14 @@ public final class XcodeBuildSystem: SPMBuildCore.BuildSystem {
         guard result.exitStatus == .terminated(code: 0) else {
             throw Diagnostics.fatalError
         }
+        
+        let stdout = try result.output.get()
+        if !stdout.isEmpty {
+            try delegate.parse(bytes: result.output.get())
```
Comment (tomerd) [inline Sources/XCBuildSupport/XcodeBuildSystem.swift:122]: can use `stdout` here?

File: Sources/XCBuildSupport/XcodeBuildSystem.swift [hunk 3]
```diff
@@ -119,6 +116,14 @@ public final class XcodeBuildSystem: SPMBuildCore.BuildSystem {
...
+        } else {
+            let stderr = try result.utf8stderrOutput()
+            self.diagnostics.emit(StringError(stderr))
```
Comment (tomerd) [inline Sources/XCBuildSupport/XcodeBuildSystem.swift:125]: should this emit "unknown error" if both stdout and stderr are empty?

File: Sources/XCBuildSupport/XcodeBuildSystem.swift (hunk 2 above)
Comment (elsh) [reply Sources/XCBuildSupport/XcodeBuildSystem.swift]: oops, updated

**Comment by tomerd** [comment]
@swift-ci smoke test

**Review by jakepetroules — CHANGES_REQUESTED** [review]

File: Sources/XCBuildSupport/XcodeBuildSystem.swift [hunk 4]
```diff
@@ -103,12 +103,7 @@ public final class XcodeBuildSystem: SPMBuildCore.BuildSystem {
 
         arguments += buildParameters.xcbuildFlags
 
-        let delegate = createBuildDelegate()
-        let redirection: Process.OutputRedirection = .stream(stdout: delegate.parse(bytes:), stderr: { bytes in
-            self.diagnostics.emit(StringError(String(bytes: bytes, encoding: .utf8)!))
-        })
-
-        let process = Process(arguments: arguments, outputRedirection: redirection)
+        let process = Process(arguments: arguments, outputRedirection: .collect)
```
[the remark this answers is not in the document]
Comment (abertelrud) [reply Sources/XCBuildSupport/XcodeBuildSystem.swift:106]: That seems like a good suggestion, Jake.

[the remark this answers is not in the document]
Comment (elsh) [reply Sources/XCBuildSupport/XcodeBuildSystem.swift:106]: thanks for the suggestion, will update

**Comment by elsh** [comment]
@swift-ci smoke test

**Review by neonichu — APPROVED** [review]

**Review by abertelrud — APPROVED** [review]
Looks good to me.  Thanks!

File: Sources/XCBuildSupport/XcodeBuildSystem.swift [hunk 5]
```diff
@@ -104,10 +104,15 @@ public final class XcodeBuildSystem: SPMBuildCore.BuildSystem {
...
+        var hasStdout = false
+        var stderrBuffer: [UInt8] = []
+        let redirection: Process.OutputRedirection = .stream(stdout: { bytes in
+            hasStdout = hasStdout || !bytes.isEmpty
+            delegate.parse(bytes: bytes)
+        }, stderr: { bytes in
+            stderrBuffer.append(contentsOf: bytes)
         })
-
+
```
Comment (tomerd) [inline Sources/XCBuildSupport/XcodeBuildSystem.swift:115]: remove empty line (format)

**Review by jakepetroules — APPROVED** [review]

**Comment by elsh** [comment]
@swift-ci smoke test

**Pull request merged**